source/source.go

```diff
@@ -109,21 +148,64 @@ func (s *Source) Open(ctx context.Context, pos sdk.Position) error {
 		Str("consumerName", *consumerResponse.Consumer.ConsumerName).
 		Msg("kinesis consumer registered")
 
+	if err := s.waitForConsumer(ctx, consumerResponse.Consumer); err != nil {
+		return fmt.Errorf("error waiting for consumer: %w", err)
+	}
+
 	s.consumerARN = consumerResponse.Consumer.ConsumerARN
 	err = s.subscribeShards(ctx, pos)
 	if err != nil {
 		return err
 	}
 
-	go s.listenEvents(ctx)
+	// listenEvents will use tomb.Go, so we initialize it here
+	s.tomb = &tomb.Tomb{}
+	s.listenEvents(ctx)
+
+	sdk.Logger(ctx).Info().Msg("source ready to be read from")
 
 	return nil
 }
 
-func (s *Source) Read(ctx context.Context) (sdk.Record, error) {
+func (s *Source) waitForConsumer(ctx context.Context, consumer *types.Consumer) error {
+	for count := 1; count <= 5; count++ {
+		describedConsumer, err := s.client.DescribeStreamConsumer(ctx, &kinesis.DescribeStreamConsumerInput{
+			ConsumerARN:  consumer.ConsumerARN,
+			ConsumerName: consumer.ConsumerName,
+			StreamARN:    &s.config.StreamARN,
+		})
+		if err != nil {
+			return fmt.Errorf("failed to describe stream consumer: %w", err)
+		}
+		if describedConsumer.ConsumerDescription.ConsumerStatus == types.ConsumerStatusActive {
+			return nil
+		}
+
+		secsToWait := math.Exp2(float64(count))
+		sdk.Logger(ctx).Info().
+			Str("consumerARN", *consumer.ConsumerARN).
+			Float64("seconds", secsToWait).
+			Msg("waiting for consumer to be ready")
+
+		select {
+		case <-ctx.Done():
+			return fmt.Errorf("context timeout while trying for consumer to be ready")
```

Comment:
If the context is done I'd expect to see the context error returned.

```suggestion
			return ctx.Err()
```

Or, if we want to include context around the error

```suggestion
			return fmt.Errorf("failed to wait for consumer to be ready: %w", ctx.Err())
```